mirrord/layer/src/outgoing/tcp.rs

```diff
@@ -213,30 +265,54 @@ impl TcpOutgoingHandler {
-                            let mirror_listener = match remote_address {
-                                SocketAddr::V4(_) => {
-                                    TcpListener::bind(SocketAddr::new(
-                                        IpAddr::V4(Ipv4Addr::UNSPECIFIED),
-                                        0,
-                                    ))
-                                    .await?
+                            // Socket held by the layer, waiting for the user application to
+                            // connect to it.
+                            let layer_socket = match remote_address {
+                                SocketAddress::Ip(SocketAddr::V4(_)) => {
+                                    let socket = Socket::new(Domain::IPV4, Type::STREAM, None)?;
+                                    socket.bind(
+                                        &(SocketAddr::new(IpAddr::V4(Ipv4Addr::UNSPECIFIED), 0)
+                                            .into()),
+                                    )?;
+                                    socket
+                                }
+                                SocketAddress::Ip(SocketAddr::V6(_)) => {
+                                    let socket = Socket::new(Domain::IPV6, Type::STREAM, None)?;
+                                    socket.bind(
+                                        &(SocketAddr::new(IpAddr::V6(Ipv6Addr::UNSPECIFIED), 0)
+                                            .into()),
+                                    )?;
+                                    socket
                                 }
-                                SocketAddr::V6(_) => {
-                                    TcpListener::bind(SocketAddr::new(
-                                        IpAddr::V6(Ipv6Addr::UNSPECIFIED),
-                                        0,
-                                    ))
-                                    .await?
+                                SocketAddress::Unix(_) => {
+                                    let socket = Socket::new(Domain::UNIX, Type::STREAM, None)?;
+                                    let tmp_dir = temp_dir().join(UNIX_STREAMS_DIRNAME);
+                                    if !tmp_dir.exists() {
+                                        create_dir_all(&tmp_dir).await?;
+                                    }
+                                    let random_string: String = rand::thread_rng()
+                                        .sample_iter(&Alphanumeric)
+                                        .take(16)
+                                        .map(char::from)
+                                        .collect();
+                                    let pathname = tmp_dir.join(random_string);
+
+                                    let addr = SockAddr::unix(pathname)?;
+                                    // TODO: should we retry with a different random string if the
+                                    //       bind fails?
```

Comment:
What do you think?
Pro - in the very unlikely event that we generated the same random string again, we don't fail.
Con - adds more code, that is very probably unneeded.

I would go for leave it this way, but I wanted to check with you when relying on probability.

Reply:
Fine with leaving it this way